armi/reactor/components/component.py

```diff
@@ -975,8 +987,20 @@ def mergeNuclidesInto(self, compToMergeWith):
             )
 
     def iterComponents(self, typeSpec=None, exact=False):
-        if self.hasFlags(typeSpec, exact):
-            yield self
+        """
+        Yield the component or its direct child components.
+
+        Notes
+        -----
+        This could probably be made recursive to allow arbitrary depth, but it's more complex
+        """
+        if self._children:
+            for child in self._children:
+                if self.hasFlags(typeSpec, exact):
+                    yield child
+        else:
+            if self.hasFlags(typeSpec, exact):
+                yield self
```

Comment:
very nice, that does seem better. I'll try your suggestion.

Reply:
ok so I tried With that suggested iterComponents change, and my `test_trisoCase.py` unit test fails with max recursion errors. If you iterate over yourself and your children you will just keep iterating over yourself... 

I tried this kind of thing for a while and just couldn't quite get it working. I'm sure it can be done... it's just hard. 

```  File "/home/nick/repos/armi/armi/reactor/composites.py", line 1256, in <listcomp>
    [c.getNumberDensity(nuc) for nuc in nucNames]
  File "/home/nick/repos/armi/armi/reactor/composites.py", line 1256, in <listcomp>
    [c.getNumberDensity(nuc) for nuc in nucNames]
  File "/home/nick/repos/armi/armi/reactor/composites.py", line 1233, in getNumberDensity
    return self.getNuclideNumberDensities([nucName])[0]
  File "/home/nick/repos/armi/armi/reactor/components/component.py", line 605, in getNuclideNumberDensities
    composites.Composite.getNuclideNumberDensities(self, nucNames),
  File "/home/nick/repos/armi/armi/reactor/composites.py", line 1255, in getNuclideNumberDensities
    [
  File "/home/nick/repos/armi/armi/reactor/composites.py", line 1256, in <listcomp>
    [c.getNumberDensity(nuc) for nuc in nucNames]
  File "/home/nick/repos/armi/armi/reactor/composites.py", line 1256, in <listcomp>
    [c.getNumberDensity(nuc) for nuc in nucNames]
  File "/home/nick/repos/armi/armi/reactor/composites.py", line 1233, in getNumberDensity
    return self.getNuclideNumberDensities([nucName])[0]
  File "/home/nick/repos/armi/armi/reactor/components/component.py", line 605, in getNuclideNumberDensities
    composites.Composite.getNuclideNumberDensities(self, nucNames),
  File "/home/nick/repos/armi/armi/reactor/composites.py", line 1244, in getNuclideNumberDensities
    [
  File "/home/nick/repos/armi/armi/reactor/composites.py", line 1245, in <listcomp>
    c.getVolume() / (c.parent.getSymmetryFactor() if c.parent else 1.0)
  File "/home/nick/repos/armi/armi/reactor/components/component.py", line 430, in getVolume
    if self.p.volume is None:
  File "/home/nick/repos/armi/armi/reactor/parameters/parameterDefinitions.py", line 293, in __get__
    return self._getter(obj)
RecursionError: maximum recursion depth exceeded
```